### ISSUE TITLE
Use generics for event target?

### ISSUE THREAD

**Issue by Daniel15** [facebook/flow#770]
`Event` and React `SyntheticEvent` are currently defined as something like:

``` js
declare class Event {
    target: EventTarget;
```

Which means that the exact type of `target` is not known

``` js
var Foo = React.createClass({
    render() {
        return (
            <input type="checkbox" onChange={this._onChange} />
        );
    },

    _onChange(evt: SyntheticEvent): void {
        console.log(evt.target.checked); // Flow error (property 'checked' not found in EventTarget)
    }
})
```

It seems like this could be solved if `Event` and `SyntheticEvent` were generic types, something like:

``` js
declare class SyntheticEvent<TElement : EventTarget> {
    target: TElement;
```

``` js
_onChange(evt: SyntheticEvent<HTMLInputElement>): void {
    console.log(evt.target.checked); // Should work since we know that evt.target is HTMLInputElement
}
```

**Comment by graue** [comment]
Any plans here? Working with forms would be much more pleasant with a fix.

**Comment by samwgoldman** [comment]
We've definitely considered this, but it trades convenience for soundness. Do you think it's worth making the model unsound?

For example,

```js
declare var elem: HTMLElement;
elem.addEventListener("mouseover", (event: Event<HTMLInputElement>) => {
  event.target.value; // event.target isn't necessarily HTMLInputElement here
});
```

To make this sound, we would need to constrain addEventListener to only accept callbacks that are compatible with the event type, which means enumerating every possible event type and defaulting to something strict.

**Comment by Daniel15** [comment]
> Do you think it's worth making the model unsound? 

Personally I think that it's worth adding, even if it makes the code unsound. I don't think it's any worse than what people would be doing today - Any event handler code today is most likely using a catchall type like `Object` or `any`, or the developer is writing their own typedef. For example, in my code I do something like this:
```
_onChange(evt: {target: {checked: boolean}}): void {
  console.log(evt.target.checked);
}
```

**Comment by jeffmo** [comment]
Also worth mentioning: We've also been chatting a bit about having type-parameter defaults (along the lines of `type Event<Target=EventTarget> = ...`).

This would let users write something like `(event: Event<>) => ...` if they want to be strict, or `(event: Event<HTMLInputElement>) => ...` if the structure of the app is such that the event is definitely always targeting an input element.

**Comment by jeffmo** [comment]
FYI, Type param defaults went out in 0.24: https://github.com/facebook/flow/blob/master/Changelog.md#v0240

**Comment by calebmer** [comment]
Closing as this will be addressed in an upcoming release for React. If someone wants to contribute this for DOM events feel free to send a PR 👍

**Comment by callumlocke** [comment]
> Closing as this will be addressed in an upcoming release for React. If someone wants to contribute this for DOM events feel free to send a PR 👍

Was this ever addressed, and how?

**Comment by Daniel15** [comment]
See the docs here: https://flow.org/en/docs/react/events/

> To type event handlers you may use the `SyntheticEvent<T>` types

**Comment by ThomasGHenry** [comment]
TL;DR for posterity.... 

```js
handleChange = (e: SyntheticEvent<HTMLInputElement>) => {
  ...
  this.setState({ checked: e.currentTarget.checked }); // <----  currentTarget is the magic
};
```
